Mathlib/Topology/Category/Profinite/Nobeling.lean

```diff
@@ -0,0 +1,1827 @@
+    exact ⟨⟨[], List.chain'_nil⟩,⟨Or.inl rfl, rfl⟩⟩
+  | cons a as ih =>
+    rw [List.map_cons, List.prod_cons]
+    intro ha
+    specialize ih (by rw [List.chain'_cons'] at ha; exact ha.2)
+    rw [Finsupp.mem_span_image_iff_total] at ih
+    simp only [Finsupp.mem_supported, Finsupp.total_apply] at ih
+    obtain ⟨c, hc, hc'⟩ := ih
+    rw [← hc']; clear hc'
+    have hmap := fun g ↦ map_finsupp_sum (LinearMap.mulLeft ℤ (e (π C (· ∈ s)) a)) c g
+    dsimp at hmap ⊢
+    split_ifs
+    · rw [hmap]
+      exact finsupp_sum_mem_span_eval _ _ ha hc
+    · ring_nf
+      rw [hmap]
+      apply Submodule.add_mem
+      · apply Submodule.neg_mem
+        exact finsupp_sum_mem_span_eval _ _ ha hc
+      · apply Submodule.finsupp_sum_mem
+        intro m hm
+        apply Submodule.smul_mem
+        apply Submodule.subset_span
+        refine ⟨m, ⟨?_, rfl⟩⟩
+        simp only [Set.mem_setOf_eq]
+        have hmas : m.val ≤ as :=
+          hc (by simpa only [Finset.mem_coe, Finsupp.mem_support_iff] using hm)
+        refine le_trans hmas ?_
+        cases as with
+        | nil => exact (List.nil_lt_cons a []).le
+        | cons b bs =>
+          apply le_of_lt
+          rw [List.chain'_cons] at ha
+          have hlex := List.lt.head bs (b :: bs) ha.1
+          exact (List.lt_iff_lex_lt _ _).mp hlex
+
+end Fin
+
+theorem fin_comap_jointlySurjective
+    (hC : IsClosed C)
+    (f : LocallyConstant C ℤ) : ∃ (s : Finset I)
+    (g : LocallyConstant (π C (· ∈ s)) ℤ), f = g.comap (ProjRestrict C (· ∈ s)) := by
+  obtain ⟨J, g, h⟩ := @Profinite.exists_locallyConstant (Finset I)ᵒᵖ _ _ _
+    (spanCone hC.isCompact) _
+    (spanCone_isLimit hC.isCompact) f
+  exact ⟨(Opposite.unop J), g, h⟩
+
+/-- The good products span all of `LocallyConstant C ℤ` if `C` is closed. -/
+theorem GoodProducts.span (hC : IsClosed C) :
+    ⊤ ≤ Submodule.span ℤ (Set.range (eval C)) := by
+  rw [span_iff_products]
+  intro f _
+  obtain ⟨K, f', rfl⟩ : ∃ K f', f = πJ C K f' := fin_comap_jointlySurjective C hC f
+  refine Submodule.span_mono ?_ <| Submodule.apply_mem_span_image_of_mem_span (πJ C K) <|
+    spanFin C K (Submodule.mem_top : f' ∈ ⊤)
+  rintro l ⟨y, ⟨m, rfl⟩, rfl⟩
+  exact ⟨m.val, eval_eq_πJ C K m.val m.prop⟩
+
+end Span
+
+/-!
+
+## Relating elements of the well-order `I` with ordinals
+
+We choose a well-ordering on `I`. This amounts to regarding `I` as an ordinal, and as such it is
+can be regarded as the set of all strictly smaller ordinals, allowing to apply ordinal induction.
+
+### Main definitions
+
+* `ord I i` is the term `i` of `I` regarded as an ordinal.
+
+* `term I ho` is a sufficiently small ordinal regarded as a term of `I`.
+
+* `contained C o` is a predicate saying that `C` is "small" enough in relation to the ordinal `o`
+  to satisfy the inductive hypothesis.
+
+* `P I` is the predicate on ordinals about linear independence of good products, which the rest of
+  this file is spent on proving by induction.
+-/
+section Ordinal
```

Comment:
I think it would be technically a little bit better to put the module docstring for a section after the start of the section. So please move `section Ordinal` just above the module docstring that documents it.

Also elsewhere.